packages/styled-components/src/utils/stylis.ts

```diff
@@ -91,17 +91,22 @@ export default function createStylisInstance(
     }
 
     middlewares.push(selfReferenceReplacementPlugin, stringify);
-
-    return serialize(
-      compile(
-        options.namespace || prefix || selector
-          ? `${
-              options.namespace ? options.namespace + ' ' : ''
-            }${prefix} ${selector} { ${flatCSS} }`
-          : flatCSS
-      ),
-      middleware(middlewares)
-    );
+    let compiled = compile(prefix || selector ? `${prefix} ${selector} { ${flatCSS} }` : flatCSS);
+
+    if (options.namespace) {
+      compiled = compiled.map(rule => {
+        // add the namespace to the start
+        rule.value = `${options.namespace} ${rule.value}`;
```

Comment:
Actually does this work appropriately with media queries? I’m not sure if we have a test for this explicitly

Reply:
I have added an extra test for a media query inside a namespaced selector. I had to write a recursive function to go down through the rules and ensure that the namespace was added inside the media query rather than before it. 

I think it works now!